🐛Fixed Analytics Visible Trigger with the Intersection Observer Polyfill

torch2424 opened the pull request:
closes #18972

This does the following:

* Creates an example in `test/manual` to display the bug in safari.
* Fixes a bug with `intersectionRatio()` to check for a divide by zero.
* Adds a mutation observer to trigger events on `class` and `style` attribute changes.

# Example Screenshot

*Look for the `imagevisible` request in the network tab*

<img width="1440" alt="screen shot 2018-10-29 at 1 30 54 pm" src="https://user-images.githubusercontent.com/1448289/47678581-682e6280-db7f-11e8-85b8-e34069d8c766.png">
<img width="720" alt="screen shot 2018-10-29 at 1 31 00 pm" src="https://user-images.githubusercontent.com/1448289/47678582-682e6280-db7f-11e8-99d0-827335350d95.png">

lannka commented on src/intersection-observer-polyfill.js:
```diff
@@ -273,6 +286,19 @@ export class IntersectionObserverPolyfill {
+          this.handleMutationObserverNotification_.bind(this)
+      );
+      this.mutationObserver_.observe(element.ownerDocument, {
+        attributeFilter: ['class', 'style', 'hidden'],
```
this filter might not cover all the visibility change

lannka commented on src/intersection-observer-polyfill.js:
```diff
@@ -384,7 +421,15 @@ export class IntersectionObserverPolyfill {
+  const largerBoxArea = larger.width * larger.height;
+
+  // Check for a divide by zero
+  if (largerBoxArea === 0) {
```
prefer ternary operator here.

jridgewell commented on src/intersection-observer-polyfill.js:
```diff
@@ -286,6 +312,9 @@ export class IntersectionObserverPolyfill {
       if (this.observeEntries_[i].element === element) {
         this.observeEntries_.splice(i, 1);
+        if (this.observeEntries_.length <= 0) {
+          this.disconnect();
```
`disconnect` isn't entirely correct here, since it's API is defined by `Service`, not the InOb Polyfill.

Please add a `disconnectMutationObserver_` method.

lannka replied:
this.disconnect() call seems not needed here

jridgewell commented on src/intersection-observer-polyfill.js:
```diff
@@ -273,6 +286,19 @@ export class IntersectionObserverPolyfill {
+      this.mutationObserver_ = new MutationObserver(
+          this.handleMutationObserverNotification_.bind(this)
+      );
+      this.mutationObserver_.observe(element.ownerDocument, {
```
Use the element's ampdoc's document, not the element's `ownerDocument`.

lannka replied:
ampdoc.getRootNode can give ShadowRoot. is it the right tree to listen for mutation events?

i.e., can a change from a sibling node of this ShadowRoot affect its visibility?

jridgewell commented on src/intersection-observer-polyfill.js:
```diff
@@ -375,16 +404,33 @@ export class IntersectionObserverPolyfill {
+   * @private
+   */
+  handleMutationObserverNotification_() {
+    this.tick(this.viewport_.getRect());
```
This is gonna fire a lot. `FixedLayer` is getting by with [checking](https://github.com/ampproject/amphtml/blob/master/src/service/fixed-layer.js#L208-L215) if a hidden attribute is changed, not _any_ attribute.

lannka replied:
visibility can be toggled by attributes other than `hidden`, for example an arbitrary attribute that has CSS selector with `display: none` on it.

torch2424 replied:
In my testing, this was only called on page load, and after all bind changes had resolved. But I'll go ahead, and add a throttle 😄

lannka replied:
@torch2424 from your experiment we see all mutation events are already batched right?
@jridgewell do you see otherwise?

jridgewell replied:
> for example an arbitrary attribute that has CSS selector with display: none on it.

Yes, but `FixedLayer`, [on-actions](https://www.ampproject.org/docs/interaction_dynamic/amp-actions-and-events#globally-defined-events-and-actions), and [amp-bind](https://www.ampproject.org/docs/reference/components/amp-bind#bindings) are taking the position that the only official way is the `hidden` attribute.

> this was only called on page load, and after all bind changes had resolved

This will be fired after every animation frame that has an attribute mutation. Like very amp-element's `buildCallback` and `layoutCallback`, etc.

> we see all mutation events are already batched right?

Correct, they're batched per animation frame.

lannka replied:
> Yes, but FixedLayer, on-actions, and amp-bind are taking the position that the only official way is the hidden attribute.

how about `attempCollapse` or even `attempChangeHeight`?

jridgewell replied:
`attemptCollapse` toggles `hidden` (so does expanding). `attemptChangeHeight` does not change the hidden visibility of an element.

lannka replied:
`attemptChangeHeight` could have changed sibling's height hence moved the target element in/out viewport.

jridgewell replied:
Correct, and there are numerous ways for CSS sibling selectors to change hidden visibility, too. The runtime's position is that only the `hidden` attribute is the proper way.

lannka replied:
not even CSS sibling selector (which might not be so common). amp-bind can dynamically change CSS classes on an element and toggle visibility. That's to say, if Fandango shopping work flow later has moved away from hidden attribute to something else, then the analytics bug would come again.

overall it's a tradeoff between visibility tracking accuracy / coverage VS runtime performance. I will let @cramforce make the call here.

cramforce replied:
Is it worth creating a nasty "DisplayObserver polyfill" like this one?
https://output.jsbin.com/texepexaxa/5/quiet

It costs 1 iframe :)

torch2424 commented on src/intersection-observer-polyfill.js:
```diff
@@ -273,6 +285,19 @@ export class IntersectionObserverPolyfill {
+    // Add a mutation observer to tick ourself
+    if (!this.mutationObserver_) {
+      this.viewport_ = Services.viewportForDoc(element);
+      this.mutationObserver_ = new MutationObserver(
```
Supported by everything 😄 

https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver#Browser_compatibility

jridgewell commented on src/intersection-observer-polyfill.js:
```diff
@@ -273,6 +284,19 @@ export class IntersectionObserverPolyfill {
 
+
+    // Add a mutation observer to tick ourself
+    if (MutationObserver && !this.mutationObserver_) {
```
This will throw an error if `MutationObserver` is not defined. Use `win.MutationObserver`

jridgewell commented on src/intersection-observer-polyfill.js:
```diff
@@ -233,12 +235,21 @@ export class IntersectionObserverPolyfill {
+     * @private {?Object}
+     */
+    this.mutationObserver_ = null;
+    /** @private {?./service/viewport/viewport-impl.Viewport} */
```
Nit: newline before this.

torch2424 replied:
Done.

erwinmombay commented on src/intersection-observer-polyfill.js:
```diff
@@ -233,12 +236,25 @@ export class IntersectionObserverPolyfill {
+     * Mutation observer to fire off on visibility changes
+     * @private {?Object}
+     */
+    this.mutationObserver_ = null;
```
why is mutationobserver just an `Object` type?

torch2424 replied:
Done.

erwinmombay commented on src/intersection-observer-polyfill.js:
```diff
@@ -273,6 +289,25 @@ export class IntersectionObserverPolyfill {
+          this.handleMutationObserverNotification_.bind(this)
+      );
+      this.mutationObserver_.observe(ampdoc.getRootNode(), {
+        attributes: true,
```
do we need these properties preserved (unmangled)?

torch2424 replied:
Discussed offline: "it should get preserved if it knows its a mutation observer instance"

torch2424 commented:
@lannka @jridgewell @erwinmombay 

Made all requested changes 😄

torch2424 commented on src/intersection-observer-polyfill.js:
```diff
@@ -375,16 +413,51 @@ export class IntersectionObserverPolyfill {
+      return;
+    }
+
+    // Wait one animation frame so that other mutations may arrive.
```
Sounds good, we can wait for the decision there 😄

lannka commented:
@cramforce Interesting trick. But seems this only detects hidden to
visible, not visible to hidden?

On Mon, Oct 29, 2018, 20:17 Malte Ubl <notifications@github.com wrote:

> *@cramforce* commented on this pull request.
> ------------------------------
>
> In src/intersection-observer-polyfill.js
> <https://github.com/ampproject/amphtml/pull/19034#discussion_r229164807>:
>
> > @@ -375,16 +404,33 @@ export class IntersectionObserverPolyfill {
>      changeEntry.target = element;
>      return changeEntry;
>    }
> +
> +  /**
> +   * Handle Mutation Oberserver events
> +   * @private
> +   */
> +  handleMutationObserverNotification_() {
> +    this.tick(this.viewport_.getRect());
>
> Is it worth creating a nasty "DisplayObserver polyfill" like this one?
> https://output.jsbin.com/texepexaxa/5/quiet
>
> It costs 1 iframe :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/19034#discussion_r229164807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIGnAcguPidEtMuppx1NPb_K0PHkfQYnks5up8SugaJpZM4YAQjk>
> .
>

cramforce commented:
@lannka I'm sure @jridgewell has a better hack. I think that otherwise limiting things to `hidden` is the only feasible way. It is incomplete, but at least possible to implement.

lannka commented:
@cramforce that SGTM if monitoring all DOM change is a big performance concern, we can limit it to "hidden" attribute only at this moment.

lannka commented on src/intersection-observer-polyfill.js:
```diff
@@ -375,16 +413,51 @@ export class IntersectionObserverPolyfill {
+    }
+
+    // Wait one animation frame so that other mutations may arrive.
+    this.mutationPass_.schedule(16);
```
@torch2424 @jridgewell if the MutationObserver API already batch events to animation frames, why do we need this delay here?

torch2424 replied:
From my testing, I noticed that Mutation Observer is batched, though I trust @jridgewell 's judgement on the issue that we should use this.

jridgewell replied:
The original design for `FixedLayer` had all "updates" happening behind the pass. This would queue up the request when a sync request came in, or when an async mutation was detected. But, there were a bunch of testing changes that would have been required, so I undid the sync changes.

I can't think of a technical reason to need this anymore.

jridgewell commented:
> SGTM if monitoring all DOM change is a big performance concern, we can limit it to "hidden" attribute only at this moment.

For `FixedLayer`, it was a requirement: `FixedLayer` causes `style` mutations to do its measurements, which would in turn queue more mutation observer events, which would cause more `style` mutations, on and on and on. `FixedLayer` does not mutate `hidden` (actually it does, but it does it out of the DOM so it won't trigger an attribute mutation), so this was easy to break.

However, there's still a huge perf hit that can happen with an any-attr observer. Attributes change all the time during the element lifecycle, and it's very likely that as a user scrolled we would start laying out elements and queueing mutations. `hidden` also fixes this.

torch2424 commented:
@lannka @jridgewell 

Made changes, added hidden to [Mutation Observer Attribute Filter](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserverInit), and removed the pass 😄

jridgewell commented on src/intersection-observer-polyfill.js:
```diff
@@ -294,16 +291,12 @@ export class IntersectionObserverPolyfill {
       );
       this.mutationObserver_.observe(ampdoc.getRootNode(), {
         attributes: true,
+        attributeList: ['hidden'],
```
Hoo, I didn't know about this. Actually, looking at [MDN](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserverInit), it should be [`attributeFilter`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserverInit/attributeFilter).

torch2424 replied:
Oops! Wrote that line at the DMV will fix now

lannka commented on src/intersection-observer-polyfill.js:
```diff
@@ -273,6 +285,21 @@ export class IntersectionObserverPolyfill {
+      this.mutationObserver_ = new ampdoc.win.MutationObserver(
+          this.handleMutationObserverNotification_.bind(this)
+      );
+      this.mutationObserver_.observe(ampdoc.getRootNode(), {
```
@jridgewell was recommending using a different doc: https://github.com/ampproject/amphtml/pull/19034#discussion_r229133003

torch2424 replied:
Thanks for catching this! 😄

lannka commented on src/intersection-observer-polyfill.js:
```diff
@@ -186,7 +187,8 @@ export class IntersectionObserverApi {
  * IntersectionObserverPolyfill will call the provided callback function with
- * the change entry.
+ * the change entry. Only Works with one document for now.
+ * TODO (@torch2424): Allow this to observe elements from multiple documents.
```
could you move this comment to the code that's causing this limitation?

torch2424 commented:
@lannka @jridgewell 

Made requested changes. Let me know what you think 😄

torch2424 commented:
@jridgewell Now I did 😄 

#19054

torch2424 commented:
Did some testing before merging, seems like the mutation observer is called before the repaint (and display of the element). Thus, we need to wait for a vsync before ticking. Will implement before merging.

torch2424 commented:
Discussed offline with @jridgewell and @lannka Agreed to just re-add the pass to fix the above mentioned issue.

Will merge once travis is done running
